### PR TITLE
SNOW-647377 Support quoted column names

### PR DESCRIPTION
**Pull request by sfc-gh-lsembera**
This PR implements support for quoted column names. In the code, internal unquoted name is used everywhere with the exception of statistics because REST API expects column display name in EpInfo.

Testing
Integration tests have been added with coverage for various corner cases.

### CODE REVIEW

File: src/main/java/net/snowflake/ingest/streaming/internal/ArrowRowBuffer.java [hunk 1]
```diff
@@ -442,7 +444,7 @@ private float convertRowToArrow(
     float rowBufferSize = 0F;
     for (Map.Entry<String, Object> entry : row.entrySet()) {
       rowBufferSize += 0.125; // 1/8 for null value bitmap
-      String columnName = formatColumnName(entry.getKey());
+      String columnName = LiteralQuoteUtils.unquoteColumnName(entry.getKey());
```
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/ArrowRowBuffer.java:447]: I don't think this is correct since we need the display name for KC to do schema evolution, for example if the input has a column called "\"c1\"" and the table doesn't have it, then we will return this extra column name in InsertValidationResponse.InsertError and KC will create the column "c1" in the table, so the wrong column will be created if you pass the unquoted name. I believe the unquotedName is only used in the Arrow scanner and all other places should be using the display name, free feel to schedule a chat with me if needed

**Review by sfc-gh-tzhang — CHANGES_REQUESTED** [review]

File: src/main/java/net/snowflake/ingest/streaming/internal/ArrowRowBuffer.java (hunk 1 above)
Comment (sfc-gh-azagrebin) [reply src/main/java/net/snowflake/ingest/streaming/internal/ArrowRowBuffer.java:447]: Maybe it is easier to do this the other way around: to use display name only for `InsertValidationResponse.InsertError`
if we send both display and unquoted name from server side?

The problem with using display name to match the user input is that we have to unquote it and remove quote and space escapes to match the server side logic and then quote it back to match against display name. and we still need it for Arrow atm.
We could write custom logic to resolve user input name against display name but then this is yet another way to do things that we do not have on server side but have to keep align.
Now we can just move `unquoteColumnName` to SF common later and reuse it in the SDK.

Comment (sfc-gh-lsembera) [reply src/main/java/net/snowflake/ingest/streaming/internal/ArrowRowBuffer.java:447]: @sfc-gh-tzhang We can discuss it in the evening, but I don't know how support this KC feature without introducing client-side quoting again. For example, if it tries to ingest into column `CREATE` and such column does not exists, KC would have to create it, but it needs to know what it is keyword and therefore quote it, right? Or is the user required to pass it quoted already?

**Review by sfc-gh-tjones — APPROVED** [review]

**Comment by sfc-gh-tzhang** [comment]
> * Extra columns in user input => Original (unquoted) user input

To make sure, we will return the original input name? For example, if the input is "CREATE", then we will return "CREATE", and if the input is "\"create\"", and we will return "\"create\""?

File: src/main/java/net/snowflake/ingest/streaming/internal/ColumnMetadata.java [hunk 2]
```diff
@@ -111,6 +111,10 @@ boolean getNullable() {
     return this.nullable;
   }
 
+  String getInternalName() {
```
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/ColumnMetadata.java:114]: Should we just store the internal name as well? Looks like there're a bunch of places are using it

File: src/main/java/net/snowflake/ingest/streaming/internal/LiteralQuoteUtils.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
+package net.snowflake.ingest.streaming.internal;
```
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/LiteralQuoteUtils.java:1]: Please add copyright

File: src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java [hunk 4]
```diff
@@ -306,20 +306,22 @@ byte[] performConversion(String input) {
   private long currentMaxLength;
   private CollationDefinition collationDefinition;
   private final String collationDefinitionString;
+  private final String columnDisplayName;
```
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java:309]: could you add a comment explain why we need it here?

**Review by sfc-gh-tzhang — APPROVED** [review]
Left some comments, otherwise LGTM, thanks!

File: src/main/java/net/snowflake/ingest/streaming/internal/ColumnMetadata.java (hunk 2 above)
Comment (sfc-gh-lsembera) [reply src/main/java/net/snowflake/ingest/streaming/internal/ColumnMetadata.java:116]: Good point, I changed it to calculate and store the internal name from name when `setName` is called. In the future, we will read it from the server.

File: src/main/java/net/snowflake/ingest/streaming/internal/LiteralQuoteUtils.java (hunk 3 above)
Comment (sfc-gh-lsembera) [reply src/main/java/net/snowflake/ingest/streaming/internal/LiteralQuoteUtils.java:4]: Added

File: src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java (hunk 4 above)
Comment (sfc-gh-lsembera) [reply src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java:310]: Comment added

**Comment by sfc-gh-lsembera** [comment]
> To make sure, we will return the original input name? For example, if the input is "CREATE", then we will return "CREATE", and if the input is ""create"", and we will return ""create""?

Yes, this is correct. In case of an extra column in the map, the validation response contains exactly what the used entered, unmodified. There is an [integration test for that](https://github.com/snowflakedb/snowflake-ingest-java/pull/293/files#diff-6e2db18062631e042dc4f6ea7a3a52d3ef5aa99f1d1465e5e476a8c4c27bcb35R118).
